sympy/logic/algorithms/lra_theory.py

```diff
@@ -0,0 +1,667 @@
+from sympy.matrices.dense import eye
+from sympy.assumptions.relation.binrel import AppliedBinaryRelation
+from sympy.assumptions.ask import Q
+from sympy.core import Dummy
+from sympy.core.mul import Mul
+from sympy.core.add import Add
+from sympy.core.relational import Eq
+from sympy import SYMPY_DEBUG
+
+
+def sep_const_coeff(expr):
+    if isinstance(expr, Add):
+        return expr, 1
+
+    if isinstance(expr, Mul):
+        coeffs = expr.args
+    else:
+        coeffs = [expr]
+
+    var, const = [], []
+    for c in coeffs:
+        if c.is_constant():
+            const.append(c)
+        else:
+            var.append(c)
+    return Mul(*var), Mul(*const)
+
+
+def list_terms(expr):
+    if not isinstance(expr, Add):
+        return [expr]
+
+    return expr.args
+
+
+def sep_const_terms(expr):
+    if isinstance(expr, Add):
+        terms = expr.args
+    else:
+        terms = [expr]
+
+    var, const = [], []
+    for t in terms:
+        if t.is_constant():
+            const.append(t)
+        else:
+            var.append(t)
+    return sum(var), sum(const)
+
+
+def standardize_binrel(prop):
+    assert prop.function in [Q.ge, Q.gt, Q.le, Q.lt, Q.eq]
+
+    expr = prop.lhs - prop.rhs
+    if prop.function in [Q.ge, Q.gt]:
+        expr = -expr
+    var, const = sep_const_terms(expr)
+
+    if prop.function == Q.eq:
+        return Q.eq(var, const)
+    if prop.function in [Q.gt, Q.lt]:
+        return Q.lt(var, const)
+    else:
+        return Q.le(var, const)
+
+
+class Boundry:
+    """
+    Represents an upper or lower bound or an equality between a symbol
+    and some constant.
+    """
+    def __init__(self, var, const, upper, equality, strict=None):
+        if not equality in [True, False]:
+            assert equality in [True, False]
+
+
+        self.var = var
+        if isinstance(const, tuple):
+            s = const[1] != 0
+            if strict:
+                assert s == strict
+            self.bound = const[0]
+            self.strict = s
+        else:
+            self.bound = const
+            self.strict = strict
+        self.upper = upper if not equality else None
+        self.equality = equality
+        self.strict = strict
+        assert self.strict is not None
+
+    def get_inequality(self):
+        if self.equality:
+            return Eq(self.var, self.bound)
+        elif self.upper and self.strict:
+            return self.var < self.bound
+        elif not self.upper and self.strict:
+            return self.var > self.bound
+        elif self.upper:
+            return self.var <= self.bound
+        else:
+            return self.var >= self.bound
+
+    def __eq__(self, other):
+        other = (other.var, other.bound, other.strict, other.upper, other.equality)
+        return (self.var, self.bound, self.strict, self.upper, self.equality) == other
+
+    def __hash__(self):
+        return hash((self.var, self.bound, self.strict, self.upper, self.equality))
+
+
+class LRASolver():
```

Comment:
I'm not sure if this is intended for inheritance or not.
If you are not sure, or experimental, I'd just try to decorate `@typing.final` to keep the users out of extending the objects.

Reply:
It's not. This specific part of the code is actually from https://github.com/sympy/sympy/pull/25363.

Reply:
I don't see what the point is of explicitly disabling subclassing like that. Especially for a class that's just for internal use anyway.

Reply:
The point is we need to discuss about ‘design’ issue before making anything public.

If we are still experimenting, or avoid design reviews, I just want to hide that as much from users.

Reply:
That's fine, but you should probably open a SYMPEP to outline any things you want the whole SymPy community to follow. Once we agree on a SYMPEP, then we can enforce/encourage the policies developed there.

Reply:
Maybe I should a leading underscore so it's _LRASolver?

Reply:
That may work